### ISSUE TITLE
Support Google Container Registry

### ISSUE THREAD

**Issue by DukeyToo** [terraform-providers/terraform-provider-google#4184]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment. If the issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If the issue is assigned to a user, that user is claiming responsibility for the issue. If the issue is assigned to "hashibot", a community member has claimed the issue already.

### Description

Right now there is no clean way to use terraform to add a private project Google Container Registry (GCR), especially in a tightly secured CI environment. GCR is backed by a bucket named `artifacts.<project>.appspot.com`, and the only way to create that bucket is by doing a `docker push` to `gcr.io/<project>`.

However, doing the `docker push` creates a resource (bucket) that is not in the known terraform state, so additional manual steps are necessary to get terraform to recognize the bucket, and apply permissions to it.

The current process:

- Temporarily grant a user or group or service account `storage.admin` role in the project. This can be done in terraform, but is only temporarily needed. 
- Manually push to gcr.io using above user/service account
- Remove the `storage.admin` role for that user/group/service account
- Manually import the newly created `artifacts.<project>.appspot.com` bucket into the terraform state file
- Add the existing bucket in terraform, and grant permissions to those that should be able to do stuff 

### New or Affected Resource(s)

* google_storage_bucket

### Potential Terraform Configuration

```tf
resource "google_storage_bucket" "artifacts" {
  name = "artifacts.my-project.appspot.com" 
  project     = "my-project"
  location = ""  # leave blank for global, or use one of the available gcr.io regions

  container_registry = true  # this is the trigger for new behavior - if set, then name must be the correct backing bucket name.
}
```

Above config would detect if the backing bucket exists, and if not, push an almost empty docker image to `gcr.io/<project>`, then update the state to include the bucket, and set any other bucket attributes, such as labels. 

### References

- https://binx.io/blog/2019/03/29/how-to-grant-access-to-the-google-container-registry/ - a decent overview of how to do this in a hacky way, by using `local-exec` in an unintended way

**Comment by slevenick** [comment]
Hey @DukeyToo 

I'm looking at how we would implement a feature like this, and I'm wondering what your use case is.

You can currently use the [docker provider](https://www.terraform.io/docs/providers/docker/index.html) to push an image to a repository to create the bucket. What else are you looking to be able to do with a terraform resource for this?

Anyone else who is interested in this feature is welcome to propose their use case as well.

**Comment by DukeyToo** [comment]
> You can currently use the [docker provider](https://www.terraform.io/docs/providers/docker/index.html) to push an image to a repository to create the bucket.

I'm not familiar with this provider, but reading the linked docs, pushing an image does not seem like a supported activity. It is certainly possible that there is a simple way to do this, but I have not found it.

My use case is to use terraform to manage permissions on the backing "artifacts" bucket, without having to do any prior manual steps such as pushing an image (to create said bucket), or importing the bucket manually into terraform state. Usually the permissions are for a service account to be able to push and/or pull images.

**Comment by slevenick** [comment]
Ah, yeah looks like pushing an image is not supported by that provider.

I've found a way to do this, but I'll have to do some investigation before putting it into a resource.

Do storage bucket IAM policy/member/bindings work for your use case to manage permissions once the GCR bucket is created? IE, this request is just to make a resource that will ensure that the bucket is created?

**Comment by DukeyToo** [comment]
Yes that is correct

**Comment by swarupdonepudi** [comment]
@slevenick how to use magic-modules to adrdress this issue?

**Comment by slevenick** [comment]
@swarupdonepudi I don't understand your question. This resource should be in the next released version of the terraform provider (3.9.0) which is expected soon
